### ISSUE TITLE
Implement: `tt_gcp_fg_setup()` and `tt_gcp_fg()`

### ISSUE THREAD

**Issue by jeremy-myers** [sandialabs/pyttb#198]
# Implementation request
`tt_gcp_fg_setup()` and `tt_gcp_fg()` are TTB functions not yet available in pyttb. They are used in MATLAB TTB to support `gcp_opt()` computations. Even without a `gcp_opt()` method, they are valuable in their own right:
* `tt_gcp_fg_setup()` provides a function handle to the desired loss/objective function and/or gradient.
* `tt_gcp_fg` computes the (full) objective function and/or gradient specified by `tt_gcp_fg_setup()`.

# Details
## `tt_gcp_fg_setup()`
Returns the function and gradient function as well as the lower bound for different types of objective functions.
### Input arguments
* `type`: `<str>`. Valid types are:
    * 'normal' or 'Gaussian'
    * 'binary' or 'Bernoulli-odds'
    * 'Bernoulli-logit'
    * 'count' or 'Poisson'
    * 'Poisson-log'
    * 'Rayleigh'
    * 'Gamma'
* `X`: input tensor. This function checks that `X` satisfies the standard constraints for that choice.

## `tt_gcp_fg()`
Computes loss function and gradient for generalized CP.
### Input arguments
* `M`: a ktensor 
* `X`: a  dense tensor.
*  `f`: a function handle of the form `f(x,m)` that measures the elementwise loss for a data entry x and corresponding model entry m.
* `g`: computes the gradient. Here `g` is the gradient of `f` and has the same form.
* `W`: specifies a weight tensor where W is a tensor that is the same size as X. It should have 1's for known values and 0's for missing values. The function/gradient is only computed w.r.t. the known values. Setting W to `None` indicates no missing data.
#### Optional arguments
* `computeF`: computes the loss function.
* `computeG` : computes only the gradient and converts it to a vector (equivalent to the `tovec()` operation on a ktensor).
* `vectorG`: computes the gradient and converts the gradient to vector form.

# Suggestion
One possible option is to implement a single wrapper to both functions that checks the constraints on `X` at compute time and returns the function and/or gradient values.

**Comment by ntjohnson1** [comment]
Good timing. I'm not sure if you saw that I updated the [gcp_opt ticket](https://github.com/sandialabs/pyttb/issues/58) but I just started scoping out the pieces for that. Thanks for the writeup here! I'm hoping to dig into this in the next few days.

**Comment by ntjohnson1** [comment]
I see you also already have a branch for this from about a month ago. If you've made more progress in the mean time also happy to collaborate if you wanted to drive this part.

**Comment by ntjohnson1** [comment]
To avoid duplication of work, erring on the side of too verbose. I linked a branch to the PR where I have all the function and gradient handles implemented. I'll assume I'm driving this.

**Comment by jeremy-myers** [comment]
@ntjohnson1 I'm happy to start from your work and add tests as I make progress. My implementation was quick-and-dirty and may not follow best practices. I see neither the branch nor the PR you mentioned (I may not have access to it).

**Comment by ntjohnson1** [comment]
My bad. That was a typo (I updated it to reflect it is on this ticket). I just wrapped up `tt_gcp_fg_setup` (and pushed to my branch). I'm trying to add a bit more hierarchy into pyttb for the gcp stuff to make it a little clearer where things are. I'm done working on this for the day, so let me know if you pick up the `tt_gcp_fg` otherwise I'll plan to finish that tomorrow and open a PR just for this piece. If there is active work you are doing that is using this or will use it feel free to try out the branch in case there is something I missed that isn't obvious prior to the full integration.

**Comment by tgkolda** [comment]
I should note that I am planning to change the way this works in Tensor Toolbox for MATLAB. The FG-evaluation should ideally be a function in/on the ktensor class. Also, the setup of the functions is separate because the setup is also used with the stochastic version of the evaluation.

**Comment by ntjohnson1** [comment]
> I should note that I am planning to change the way this works in Tensor Toolbox for MATLAB. The FG-evaluation should ideally be a function in/on the ktensor class. Also, the setup of the functions is separate because the setup is also used with the stochastic version of the evaluation.

Thanks for the heads up!

Is this landing imminently or generally in the future? My goal was to land gcp_opt support by end of next week. Unless you already have a write-up or branch with the direction you are thinking I will probably just do the implementation/translation against 3.5 then can revisit after your changes go in.

**Comment by tgkolda** [comment]
Won't be done within a week. Haha! This is really fantastic that this will be working so soon.
